libs/server/Storage/Functions/MainStore/PrivateMethods.cs

```diff
@@ -525,17 +553,34 @@ static void CopyValueLengthToOutput(ref SpanByte value, ref SpanByteAndMemory ou
         /// a. ConcurrentWriter
         /// b. PostSingleWriter
         /// </summary>
-        void WriteLogUpsert(ref SpanByte key, ref SpanByte input, ref SpanByte value, long version, int sessionID)
+        void WriteLogUpsert(ref SpanByte key, ref RawStringInput input, ref SpanByte value, long version, int sessionId)
         {
             if (functionsState.StoredProcMode) return;
 
             //We need this check because when we ingest records from the primary
             //if the input is zero then input overlaps with value so any update to RespInputHeader->flags
             //will incorrectly modify the total length of value.
             if (input.Length > 0)
-                ((RespInputHeader*)input.ToPointer())->flags |= RespInputFlags.Deterministic;
+                input.header.flags |= RespInputFlags.Deterministic;
+
+            // Serializing key, RawStringInput & value to RMW log
+            var parseStateArgCount = input.parseState.Count - input.parseStateStartIdx;
 
-            functionsState.appendOnlyFile.Enqueue(new AofHeader { opType = AofEntryType.StoreUpsert, version = version, sessionID = sessionID }, ref key, ref input, ref value, out _);
+            var sbToSerialize = new SpanByte[3 + parseStateArgCount];
```

Comment:
This is a critical path allocation. We should instead give the parse state to Enqueue which should natively handle and serialize it directly into TsavoriteLog - this holds for all instances of log enqueue in the main and object store.